Chore: test simplification for stateful set

luoqr96 opened the pull request:
simplify the tests that involved with stateful set in pkg/apis/apps/validation/validation_test.go

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind cleanup

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

Table-driven tests are awesome, but they can be very verbose and obscure what's actually being tested in each case.

This PR simplifies the usage of stateful set in pkg/apis/apps/validation/validation_test.go

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->

This PR only solves a small part of  #114511

#### Special notes for your reviewer:

Thank you for taking the time to review my first issue on GitHub! As this is my first time submitting an issue, I would greatly appreciate any feedback you may have on how I can improve my submission in the future.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

luoqr96 commented:
/easycla

luoqr96 commented:
/easycla

luoqr96 commented:
/easycla

thockin commented on pkg/apis/apps/validation/validation_test.go:
```diff
@@ -42,6 +42,96 @@ func intStrAddr(intOrStr intstr.IntOrString) *intstr.IntOrString {
 
+type setTweak func(s apps.StatefulSet)
+
+func makeSet(selectorLabels *map[string]string, template *api.PodTemplate, tweaks ...setTweak) apps.StatefulSet {
```
`makeStatefulSet` or `mkStatefulSet` seem better - "Set" could be a replicase, daemonset, statefulset, ...

luoqr96 replied:
modified as suggested

thockin commented on pkg/apis/apps/validation/validation_test.go:
```diff
@@ -42,6 +42,96 @@ func intStrAddr(intOrStr intstr.IntOrString) *intstr.IntOrString {
 	return &intOrStr
 }
 
+type setTweak func(s apps.StatefulSet)
```
maybe `tweakStatefulSet` ?

luoqr96 replied:
modified as suggested

thockin commented on pkg/apis/apps/validation/validation_test.go:
```diff
@@ -42,6 +42,96 @@ func intStrAddr(intOrStr intstr.IntOrString) *intstr.IntOrString {
 
+type setTweak func(s apps.StatefulSet)
+
+func makeSet(selectorLabels *map[string]string, template *api.PodTemplate, tweaks ...setTweak) apps.StatefulSet {
```
I don't think you need a pointer to the map - just passing the map "by value" is sufficient.

But also, you could make a tweak func?  Set a sane default in the `makeStatefulSet` func, and call the tweak ONLY from the cases that need a specific selector (which emphasizes that the selector is important) like:

```
fun setSelector(labels map[string]string) tweakStatefulSet {
    return func(ss apps.StatefulSet) {
        ss.Spec.Selector = &metav1.LabelSelector{MatchLabels: &labels)}
    }
}
```

luoqr96 replied:
modified as suggested

thockin commented on pkg/apis/apps/validation/validation_test.go:
```diff
@@ -42,6 +42,96 @@ func intStrAddr(intOrStr intstr.IntOrString) *intstr.IntOrString {
+		ObjectMeta: metav1.ObjectMeta{Name: "abc", Namespace: metav1.NamespaceDefault},
+		Spec: apps.StatefulSetSpec{
+			PodManagementPolicy: apps.OrderedReadyPodManagement,
+			Selector:            &metav1.LabelSelector{MatchLabels: *selectorLabels},
```
This should produce a fully validate-safe object - I feel like you're maybe missing fields?  Isn't replicas required?  I see it missing in some test cases, so maybe not.

luoqr96 replied:
Replicas occur in 17/74 stateful set test cases, so I think maybe it is better to create a tweak func for it other than set a default value.

thockin commented on pkg/apis/apps/validation/validation_test.go:
```diff
@@ -463,118 +467,49 @@ func TestValidateStatefulSet(t *testing.T) {
-						}()},
-				},
-			},
+			set:  makeSet(&validLabels, &validPodTemplate, tweakRepicas(3), tweakUpdateStrategy(makeUpdateStrategy(tweakUpdateStrategyType(apps.OnDeleteStatefulSetStrategyType), tweakRollingUpdatePartition(1)))),
```
You might want to linewrap these cases of multiple tweaks, like:

```
set:  makeSet(&validLabels, &validPodTemplate,
     tweakRepicas(3), 
tweakUpdateStrategy(
    makeUpdateStrategy(
        tweakUpdateStrategyType(apps.OnDeleteStatefulSetStrategyType),
        tweakRollingUpdatePartition(1)
))),
```

or something - the goal is to call out what makes this case unique

luoqr96 replied:
modified as suggested

thockin reviewed (commented):
What a great cleanup!

luoqr96 commented:
Thank you! I'll simplify all the usage of stateful set in validation_test.go ASAP> What a great cleanup!

luoqr96 commented:
@thockin Modified as you suggested, please review again!

I simplified all the test cases of the stateful set in this new commit, PTAL :)

thockin commented on pkg/apis/apps/validation/validation_test.go:
```diff
@@ -62,53 +62,104 @@ func makeSet(selectorLabels *map[string]string, template *api.PodTemplate, tweak
-func tweakPodManagementPolicy(policy apps.PodManagementPolicyType) setTweak {
-	return func(s apps.StatefulSet) {
-		s.Spec.PodManagementPolicy = policy
+func tweakFinalizers(finalizers []string) statefulSetTweak {
```
Optional:

For finalizers, labels, annotations, you accept maps or slices, which makes the caller have to specify them.  If the goal is to make the callers as obvious and readable as possible, there's another way to approach this.  YOu could do:

```
func tweakFinalizers(finalizers ...string) statefulSetTweak {
```

This let's the caller say `tweakFinalizers("finalizer1")` or, if they really need more than one, `tweakFinalizers("one", "two", "three")`.

For the maps, you could do either:

```
func tweakLabels(key, value) statefulSetTweak {
    return func(ss apps.StatefulSet) {
        if ss.Labels == nil {
            ss.Labels = map[string]string{}
        }
        ss.Labels[key] = value
    }
}
```

If we really need to set more than one, you can call this more than one time.  Or you can do something like:

```
func tweakMetaLabels(keysAndValues ...string) statefulSetTweak {
```

But this is sort of awkward, IMO

luoqr96 replied:
modified as suggested.

However, as the selector labels are typically already specified, I find it maybe more convenient to continue accepting a map in the tweak function.

thockin replied:
OK - the goal is "as obvious as possible, and not one step further".  If you think that's the right tradeoff, go for it.

thockin commented on pkg/apis/apps/validation/validation_test.go:
```diff
@@ -62,53 +62,104 @@ func makeSet(selectorLabels *map[string]string, template *api.PodTemplate, tweak
+	}
+}
+
+func tweakRestartPolicy(rp api.RestartPolicy) statefulSetTweak {
```
I'd include the word Template in the name here

luoqr96 replied:
modified as suggested

thockin commented on pkg/apis/apps/validation/validation_test.go:
```diff
@@ -62,53 +62,104 @@ func makeSet(selectorLabels *map[string]string, template *api.PodTemplate, tweak
+func mkUpdateStrategy(tweaks ...updateStrategyTweak) apps.StatefulSetUpdateStrategy {
 	s := apps.StatefulSetUpdateStrategy{
 		Type: apps.RollingUpdateStatefulSetStrategyType,
+		RollingUpdate: func() *apps.RollingUpdateStatefulSetStrategy {
```
Why this `func` construct?  I'm missing the purpose of it.

luoqr96 replied:
removed

thockin commented on pkg/apis/apps/validation/validation_test.go:
```diff
@@ -366,267 +364,169 @@ func TestValidateStatefulSet(t *testing.T) {
-				},
-			},
+			set: mkStatefulSet(&validPodTemplate,
+				tweakReplicas(3),
```
This is an awesome example - is Replicas ACTUALLY important to this test case, or was it just set for no real reason?  It doesn't seem like it SHOULD matter.

This cleanup makes such things obvious, and now there can be followups to remove unnecessary noise  like this (if it is unnecessary :).  Followups, please - not the same PR.

thockin reviewed (commented):
A few small comments.

This PR is now almost too big to review.  My brsain gets fatigued and will miss things at this size.  Ideally this would be a series of small commits:
1) add the basic "make" function and a couple of simple tweaks, use that to show the pattern
2...N) add one more tweak and use it in those cases that can now be converted (occasionally you have to add 2 tweaks to satisfy a case)
last) fix any obviously redundant or wrongly named cases 

Another asspect of this you might consider as a followup is adding the line-number to each test case (`line string // if not empty, will be logged with errors, use line() to set`) and a line() function:

```
// line returns the line number of the caller, if possible.  This is useful in
// tests with a large number of cases - when something goes wrong you can find
// which case more easily.
func line() string {                                                                                                                        
    _, _, line, ok := stdruntime.Caller(1)
    var s string
    if ok {
        s = fmt.Sprintf("%d", line) 
    } else {
        s = "<??>"
    }
    return s
}
```

And then print that in the tests  (e.g. in the "Run()" call).

See pkg/registry/core/service/storage/storage_test.go for example.  This can also be a followup or at least a different commit.

FINALLY (again as followup or as a final commit, all on its own) you can do whitespace cleanups, so things like:

```
	successCases := []testCase{
		{
			name:   "update replica count",
			old:    mkStatefulSet(&validPodTemplate),
			update: mkStatefulSet(&validPodTemplate, tweakReplicas(3)),
		},
		{
			name:   "update containers 1",
			old:    mkStatefulSet(&validPodTemplate),
			update: mkStatefulSet(addContainersValidTemplate),
		},
		{
			/// and so on
		},
	}
```

become:

```
	successCases := []testCase{{
		name:   "update replica count",
		old:    mkStatefulSet(&validPodTemplate),
		update: mkStatefulSet(&validPodTemplate, tweakReplicas(3)),
	}, {
		name:   "update containers 1",
		old:    mkStatefulSet(&validPodTemplate),
		update: mkStatefulSet(addContainersValidTemplate),
	}, {
		/// and so on
	}}
```

10K lines of tests will become 1K or less.  It's like magic.

luoqr96 committed:
Chore: add basic make function for stateful set

Add name and namespace tweak functions

luoqr96 committed:
Chore: add pod management policy tweak function

luoqr96 committed:
Chore: add replicas tweak function

luoqr96 committed:
Chore: add template restart policy tweak function

luoqr96 committed:
Chore: add min ready seconds tweak function

luoqr96 committed:
Chore: add ordinal start tweak function

luoqr96 committed:
Chore: add finalizers tweak function

luoqr96 committed:
Chore: add annotation tweak function

luoqr96 committed:
Chore: add labels tweak function

luoqr96 committed:
Chore: add update strategy type tweak function

luoqr96 committed:
Chore: add rolling update partition type tweak function

luoqr96 committed:
Chore: add rolling update max unavailable type tweak function

luoqr96 committed:
Chore: add persistent volume claim template tweak function

luoqr96 committed:
Chore: add persistent volume claim retention policy tweak function

luoqr96 committed:
Chore: add selector labels tweak function

luoqr96 commented:
Hi Hockin! I reorganized all the commits as suggested, PTAL!

I apologize for making a mess before.

@thockin

thockin commented on pkg/apis/apps/validation/validation_test.go:
```diff
@@ -944,24 +928,12 @@ func TestValidateStatefulSetUpdate(t *testing.T) {
+			old: mkStatefulSet(&validPodTemplate,
+				tweakSelectorLabels(validLabels),
+			),
+			update: mkStatefulSet(&validPodTemplate2,
```
single tweaks probably don't need linewraps.

luoqr96 replied:
modified as suggested

thockin commented on pkg/apis/apps/validation/validation_test.go:
```diff
@@ -42,6 +42,38 @@ func intStrAddr(intOrStr intstr.IntOrString) *intstr.IntOrString {
+	}
+
+	for _, tw := range tweaks {
+		tw(s)
```
This needs to be a pointer - I see you did that in the last commit, but each of these commits, as beautiful as they are, don't work?  Can you back-populate that into these commits?

luoqr96 commented on pkg/apis/apps/validation/validation_test.go:
```diff
@@ -507,48 +504,27 @@ func TestValidateStatefulSet(t *testing.T) {
-			},
+			set: mkStatefulSet(&validPodTemplate,
+				tweakPodManagementPolicy("foo"),
+				tweakReplicas(3),
```
Yes, I keep the replicas here as a literal translation. Would it be better to remove these unnecessary tweaks in a new commit or a new pull request?

thockin replied:
new PR is fine.  Let's make this PR "done" now and start a new one

luoqr96 commented on pkg/apis/apps/validation/validation_test.go:
```diff
@@ -315,21 +324,9 @@ func TestValidateStatefulSet(t *testing.T) {
-				},
-			},
+			set: mkStatefulSet(&validPodTemplate,
+				tweakMetaLabels("NoUppercaseOrSpecialCharsLike=Equals", "bar"),
```
modified as suggested

luoqr96 commented:
Hi Hockin!

 I wanted to let you know that I have addressed all of the code comments you made on this pull request. I would appreciate it if you could take another look at the code to confirm that the changes meet your expectations. Thank you for your time and feedback.

By the way, I am not sure whether it would be better to remove unnecessary tweaks in a new commit or a new pull request. For now, I have left them unmodified as a literal translation.

@thockin

thockin commented:
Thanks!

/lgtm
/approve

luoqr96 commented:
/retest

luoqr96 commented:
Hi Hockin!

The CI shows the kubectl_request_timeout test failed. I pasted the log below.

```
[31mFAIL!
[5651](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/116722/pull-kubernetes-integration/1638399925064568832#1:build-log.txt%3A5651)
(Bmessage:NAME        READY   STATUS    RESTARTS   AGE
valid-pod   0/1     Pending   0          1s
I0322 04:48:53.102939   38649 streamwatcher.go:114] Unable to decode an event from the watch stream: context deadline exceeded
has not:Timeout
42 /home/prow/go/src/k8s.io/kubernetes/test/cmd/../../test/cmd/request-timeout.sh
!!! [0322 04:48:53] Call tree:
!!! [0322 04:48:53]  1: /home/prow/go/src/k8s.io/kubernetes/test/cmd/../../third_party/forked/shell2junit/sh2ju.sh:47 run_kubectl_request_timeout_tests(...)
!!! [0322 04:48:53]  2: /home/prow/go/src/k8s.io/kubernetes/test/cmd/../../third_party/forked/shell2junit/sh2ju.sh:112 eVal(...)
!!! [0322 04:48:53]  3: /home/prow/go/src/k8s.io/kubernetes/test/cmd/legacy-script.sh:141 juLog(...)
!!! [0322 04:48:53]  4: /home/prow/go/src/k8s.io/kubernetes/test/cmd/legacy-script.sh:624 record_command(...)
!!! [0322 04:48:53]  5: hack/make-rules/test-cmd.sh:194 runTests(...)
```

However, this test passed in my local environment. I also pasted the corresponding logs below.

```
Successful
message:NAME        READY   STATUS    RESTARTS   AGE
valid-pod   0/1     Pending   0          0s
I0322 14:10:25.249865   45141 streamwatcher.go:114] Unable to decode an event from the watch stream: context deadline exceeded (Client.Timeout or context cancellation while reading body)
has:Timeout
```

Considering this PR is just a cleanup of pkg/apis/apps/validation/validation_test.go, I think the failure here is unrelated with this PR.

Could you please help me rerun this failed test? Or is there any bugs that I might have missed?

Thank you so much!

@thockin

luoqr96 commented:
/retest

pacoxu commented:
/ok-to-test
/test pull-kubernetes-integration

luoqr96 commented:
Hi Hockin! I noticed that the test freeze has just started. Should I leave this PR as is and wait for the v1.28 release cycle?

@thockin

thockin commented:
Yeah, it can't merge until the 1.28 window opens, but that's OK.  It's ready to go and it seems unlikely to grow a conflict in this area.
